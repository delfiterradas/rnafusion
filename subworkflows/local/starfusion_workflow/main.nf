include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_FOR_STARFUSION       }   from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_INDEX as SAMTOOLS_INDEX_FOR_STARFUSION_CRAM  }   from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_VIEW as SAMTOOLS_VIEW_FOR_STARFUSION         }   from '../../../modules/nf-core/samtools/view/main'
include { STARFUSION                                            }   from '../../../modules/local/starfusion/detect/main'
include { CTATSPLICING_WORKFLOW                                 }   from '../ctatsplicing_workflow'

workflow STARFUSION_WORKFLOW {
    take:
        reads               // channel: [ meta, bam ]
        junctions           // channel: [ meta, junctions ]
        ch_starfusion_ref
        starfusion_fusions

    main:
        def ch_versions = Channel.empty()
        def ch_starfusion_fusions = Channel.empty()

        if (starfusion_fusions){
            ch_starfusion_fusions = reads.combine(Channel.value(file(starfusion_fusions, checkIfExists:true)))
                                    .map { it -> [ it[0], it[2] ] }
        } else {
            reads_junction = reads.join(junctions)  // TODO: This join is not needed as STARFUSION can simply read from the junction file: https://github.com/STAR-Fusion/STAR-Fusion/wiki#alternatively-kickstart-mode-running-star-yourself-and-then-running-star-fusion-using-the-existing-outputs

            STARFUSION( reads_junction, ch_starfusion_ref.map { it -> it[1] })
            ch_versions = ch_versions.mix(STARFUSION.out.versions)
            ch_starfusion_fusions = STARFUSION.out.fusions

        }
    emit:
        fusions               = ch_starfusion_fusions
        versions              = ch_versions
    }

