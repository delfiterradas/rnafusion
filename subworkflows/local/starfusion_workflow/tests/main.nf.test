nextflow_workflow {

    name "Test Subworkflow STARFUSION_WORKFLOW"
    script "../main.nf"
    workflow "STARFUSION_WORKFLOW"
    tag "subworkflow"
    tag "star"
    tag "star/genomegenerate"
    tag "star/align"


    test("STARFUSION_WORKFLOW - Homo sapiens - FASTQs minigenome") {
        config './nextflow.config'

        setup {
            run("STAR_GENOMEGENERATE") {
                script "../../../../modules/nf-core/star/genomegenerate/main.nf"
                process {
                """
                    // FASTA
                    input[0] = Channel.fromPath(
                            "https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.fa", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}

                    // GTF
                    input[1] = Channel.fromPath(
                            "https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.gtf", checkIfExists: true
                        )
                        .map{ [[id: it.getName() ], it ]}
                    """
                }
            }

            run("STARFUSION_BUILD") {
            script "../../../../modules/local/starfusion/build/main.nf"
            process {
                """
                input[0] = [
                    [ id:'minigenome fasta' ],
                    file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.fa")
                ]
                input[1] = [
                    [ id:'minigenome gtf' ],
                    file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.gtf")
                ]

                input [2] = file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/CTAT_HumanFusionLib.mini.dat.gz")
                input [3] = "human"
                """
            }
        }
        }

        when {
            workflow {
                """
                // ch_reads
                input[0] = Channel.of(
                    [
                        [ id: "test_fastqs" ],
                        [
                            file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_1.fastq.gz", checkIfExists: true),
                            file("https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/v0.0.1/rnaseq_2.fastq.gz", checkIfExists: true)
                        ]
                    ] )

                // ch_gtf
                input[1] =
                    Channel.fromPath(
                        "https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.gtf", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_starindex_ref
                input[2] = STAR_GENOMEGENERATE.out.index

                // ch_fasta
                input[3] =
                    Channel.fromPath(
                        "https://github.com/STAR-Fusion/STAR-Fusion-Tutorial/raw/master/minigenome.fa", checkIfExists: true
                    )
                    .map{ [ [ id: it.name ], it ] }

                // ch_starfusion_ref
                input[4] = STARFUSION_BUILD.out.reference

                """
            }
            params {
                starfusion = true
                cram = 'starfusion'
                outdir = "$outputDir"
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out.fusions,
                    file(workflow.out.star_stats[0][1]).name,
                    workflow.out.star_gene_count,
                    workflow.out.ch_bam_sorted,
                    workflow.out.ch_bam_sorted_indexed,
                    workflow.out.versions
                ).match() }
            )
        }
    }




}
