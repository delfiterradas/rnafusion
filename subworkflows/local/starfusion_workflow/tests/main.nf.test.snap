{
    "STARFUSION_WORKFLOW - Homo sapiens - FASTQs minigenome": {
        "content": [
            [
                [
                    {
                        "id": "test_fastqs"
                    },
                    "test_fastqs.starfusion.fusion_predictions.tsv:md5,abe17134a231642edf9351e4964e8a97"
                ]
            ],
            "test_fastqs.Log.final.out",
            [
                [
                    {
                        "id": "test_fastqs"
                    },
                    "test_fastqs.ReadsPerGene.out.tab:md5,8e0d42deeea09924d5c7ba3147bbfd78"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,afde4f2fd6056df81e322b3c35ab7a8a",
                "versions.yml:md5,e53f1ec32bc78a33f99892e42274833a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2024-12-23T18:25:52.838411738"
    }
}